lib/postgres/vacuum/jobs/monitor_job.rb

```diff
@@ -98,14 +98,41 @@ def with_each_db_name_and_connection
           databases = Set.new
           ActiveRecord::Base.connection_handler.connection_pools.map do |connection_pool|
             db_name = connection_pool.db_config.configuration_hash[:database]
+            next unless databases.add?(db_name)
+
+            # ActiveRecord allocates a connection pool per call to `.establish_connection`
+            # As a result, multiple pools might interact with the same database, so we use
+            # the database name to dedupe.
+            connection_pool.with_connection do |connection|
+              original_timeout = statement_timeout(connection)
+              set_statement_timeout(connection, "#{configured_timeout_seconds}s")
+
+              yield(db_name, connection)
+            ensure
+              set_statement_timeout(connection, original_timeout)
+            end
 
-            # activerecord allocates a connection pool per call to establish_connection
-            # multiple pools might interact with the same database so we use the
-            # database name to dedup
-            connection_pool.with_connection { |conn| yield(db_name, conn) } if databases.add?(db_name)
+          # We want to avoid hanging onto a bad connection that would cause all future
+          # jobs to fail, so we eagerly clear the pool.
+          rescue ActiveRecord::StatementInvalid
```

Comment:
I take it that since we're now eagerly clearing the pool, we wouldn't expect to need to handle the `ActiveRecord::ConnectionTimeoutError` we've seen as a result of not clearing the pool?

Reply:
That was my hope that cleaning the pool could help avoid `ConnectionTimeoutError`, but seems like a good idea to clean the pool if we get `ConnectionTimeoutError` too. Updated.